File: Graph(그래프)/백준_1697_숨바꼭질/숨바꼭질.java
import java.util.*;

public class Main {
    static int a, b;
    static Queue<Integer> q = new LinkedList();
    static boolean[] check = new boolean[100001];
    static int[] seconds = new int[100001];
    static ArrayList<Integer> _move = new ArrayList();

    static void BFS(int n) {
        q.add(n);
        check[n] = true;
        while (q.isEmpty() == false) {
            int f = q.peek();
            q.poll();
            _move.add(f - 1);
            if (f < b) {
                _move.add(f + 1);
                _move.add(f * 2);
            }
            for (int i : _move) {
                if (i >= 0 && i <= 100000) {
                    if (check[i] == false) {
                        check[i] = true;
                        q.add(i);
                        seconds[i] = seconds[f] + 1;
                        if (i == b)
                            return;
                    }
                }
            }
            _move.clear();
        }
    }

    public static void main(String[] args) {
        Scanner s = new Scanner(System.in);
        a = s.nextInt();
        b = s.nextInt();

        BFS(a);

        System.out.println(seconds[b]);

    }
}

File: Graph(그래프)/백준_1697_숨바꼭질/숨바꼭질.cpp
#include <iostream>
#include <vector>
#include <queue>
using namespace std;

int a, b;
queue<int> q;
vector<int> _move;
bool check[100001];
int seconds[100001];

void BFS(int n)
{
    q.push(n);
    check[n] = true;

    while (q.empty() == false)
    {
        int f = q.front();
        q.pop();
        _move.push_back(f - 1);
        if (f < b) // 목적지 보다 클때 굳이 더 크게 갈필요 없음.
        {
            _move.push_back(f + 1);
            _move.push_back(f * 2);
        }
        for (int i : _move)
        {
            if (i >= 0 && i <= 100000) //조사 범위 한정
            {
                if (check[i] == false)
                {
                    q.push(i);
                    check[i] = true;
                    seconds[i] = seconds[f] + 1;
                    if (i == b)
                        return;
                }
            }
        }
        _move.clear();
    }
}

int main()
{
    scanf("%d %d", &a, &b);

    BFS(a);

    cout << seconds[b];
}
